### PR TITLE
fix(logs): reject unsupported CloudWatch offsets

### LINKED ISSUE #1998: [Bug]: /api/logs/search advertises offset pagination the CloudWatch provider ignores
### What's the bug?

`GET /api/logs/search` advertises offset pagination through `paginatedResponse`, but whether that contract holds depends on which log provider is configured. The two disagree, and a client cannot tell which one it is talking to.

**Local provider** (`local.provider.ts:253`) does the right thing:

```ts
return {
  logs: results.slice(offset, offset + limit),
  total: results.length,          // the real match count
};
```

**CloudWatch provider** (`cloudwatch.provider.ts:425`) takes the parameter and drops it:

```ts
async searchLogs(
  query: string,
  sourceName?: string,
  limit: number = 100,
  _offset = 0 // CloudWatch doesn't support offset-based pagination
)
...
return {
  logs: mapped,
  total: mapped.length,           // the size of the page just built
};
```

The route passes `offset` through and hands both values to `paginatedResponse`, which builds `Content-Range: <offset>-<end>/<total>` and picks `206` vs `200` from `data.length < total`.

### Why it matters

On CloudWatch, a client walking pages the way the `Content-Range` header tells it to gets wrong results in both directions:

- **Repeats.** Requesting `offset=100` returns the same first page as `offset=0`, because the offset never reaches the query.
- **Stops early.** `total` equals the page length, so `data.length < total` is false and the response is `200` with `Content-Range: 100-199/100`. A client that pages until it has seen `total` rows stops after one page, and the range header is self-contradictory: it claims to start at 100 of a 100-row set.

Neither failure is visible to the caller. There is no field in the response saying which provider answered.

### Reproduction

With `LOG_PROVIDER` set to CloudWatch:

```
curl -H "Authorization: Bearer <admin>" "$BASE/api/logs/search?q=error&limit=10&offset=0"
curl -H "Authorization: Bearer <admin>" "$BASE/api/logs/search?q=error&limit=10&offset=10"
```

Both return the same ten entries. The second reports `Content-Range: 10-19/10`.

### Expected behavior

Either the endpoint paginates consistently across providers, or it stops advertising a contract one provider cannot honour.

### Options

1. **Emulate offset in the CloudWatch provider.** Insights has no `OFFSET`, but the query already applies `limit`; fetching `offset + limit` and slicing locally would make small offsets behave, at the cost of refetching. It does not scale to deep offsets.
2. **Switch this endpoint to a cursor.** The per-source endpoint already uses `before_timestamp`, which suits CloudWatch's model and would be consistent with `GET /api/logs/{source}`.
3. **Reject a non-zero `offset` when the provider cannot honour it**, with a coded 400, so the failure is loud rather than silent.

My preference is 3 as an immediate fix and 2 as the real one, but this is a design call rather than a bug with one obvious repair, which is why it is filed rather than folded into #1997.

### Context

Found while making `openapi/logs.yaml` match the router in #1997. That PR documents the divergence so the spec stops promising something that does not hold, but deliberately does not change behaviour. The note it adds should be removed if this is fixed.

Verified against `268d79d`.

### PR DESCRIPTION
**Pull request by Gautam-aman**
## Summary

Fixes #1998

The `/api/logs/search` endpoint advertises offset-based pagination, but the CloudWatch log provider currently ignores non-zero offsets. This can cause repeated pages and incorrect `Content-Range` headers.

This change makes the limitation explicit by rejecting non-zero offsets when the CloudWatch provider is used.

## Changes

- Reject `offset > 0` in `CloudWatchProvider.searchLogs()`.
- Return a `400` `INVALID_INPUT` error instead of silently ignoring the offset.
- Keep `offset=0` behavior unchanged.
- Add a regression test covering non-zero CloudWatch offsets.

## Why

CloudWatch Logs Insights does not support offset-based pagination. Silently ignoring the offset causes requests such as:

`offset=0&limit=10`
`offset=10&limit=10`

to return the same page while the response advertises misleading pagination metadata.

Rejecting unsupported offsets makes the API behavior explicit and prevents clients from receiving incorrect pagination results.

Cursor-based pagination can be considered separately as a longer-term solution.

## Testing

- [x] CloudWatch provider unit tests
- [x] TypeScript typecheck
- [x] ESLint
- [x] Prettier
- [x] Regression test for non-zero offsets

## Issue

Closes #1998

---
## Summary by cubic
Rejects unsupported CloudWatch offsets in `/api/logs/search` to prevent misleading pagination. Previously we ignored non‑zero offsets and returned the first page; now we return `400 INVALID_INPUT` for any non‑zero or non‑finite `offset`. `offset=0` is unchanged; other providers are unaffected.

- Migration
  - Do not send `offset` other than `0` for CloudWatch. Use `offset=0` with `limit` and adjust the time range for subsequent pages.
  - Handle `400 INVALID_INPUT` by disabling offset-based pagination for CloudWatch.

- Review notes
  - Adds an early guard in `CloudWatchProvider.searchLogs()` that rejects non‑zero or non‑finite offsets with `AppError(400, INVALID_INPUT)`.
  - Adds a regression test covering offsets `10`, `-10`, `NaN`, `Infinity`, and `-Infinity`.

<sup>Written for commit 67af63f849523bcbc03308eebd7f56ff17292342. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1999?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - CloudWatch log searches now clearly reject unsupported nonzero or invalid pagination offsets with an informative input error.
  - Preserved existing behavior for searches starting at the first page.

### CODE REVIEW

File: backend/src/providers/logs/cloudwatch.provider.ts [hunk 1]
```diff
@@ -426,11 +426,18 @@ export class CloudWatchProvider extends BaseLogProvider {
...
     logs: (LogSchema & { source: string })[];
     total: number;
   }> {
+    if (offset > 0) {
+      throw new AppError(
+        'CloudWatch log search does not support offset-based pagination',
+        400,
+        ERROR_CODES.INVALID_INPUT
+      );
+    }
```
[the remark this answers is not in the document]
Comment (Gautam-aman) [reply backend/src/providers/logs/cloudwatch.provider.ts:440]: Good catch. I was only rejecting positive offsets, which allowed negative and NaN values through. I've changed the validation to offset !== 0 and added regression coverage for negative and NaN offsets

**Review by jwfing — CHANGES_REQUESTED** [review]
## Summary

Rejecting non-zero `offset` in `CloudWatchProvider.searchLogs()` is the right immediate fix and matches option 3 that issue #1998 explicitly endorsed, but the PR does not update `openapi/logs.yaml`, which now (since #1997 merged into `main`) documents the exact opposite behaviour.

## Requirements context

This repo has no `/docs/superpowers/` directory (specs/plans live elsewhere; `docs/` is the Mintlify user-facing site), so **no matching spec/plan was found** — I assessed against issue #1998, the PR description, and the OpenAPI contract merged by #1997.

Issue #1998 lists three options and states: *"My preference is 3 as an immediate fix and 2 as the real one"*, i.e. reject non-zero `offset` with a coded 400 now, cursor pagination later. The implementation follows that, and the cursor redesign is correctly out of scope. The issue also states, about the note #1997 added to the OpenAPI spec: *"The note it adds should be removed if this is fixed."* — that part is missing here.

Verified call-path blast radius: `searchLogs` is reached only from `backend/src/services/logs/log.service.ts:251` ← `backend/src/api/routes/logs/index.routes.ts:129`; the dashboard's `LogService.searchLogs` (`packages/dashboard/src/features/logs/services/log.service.ts:50`) has no callers today (`useLogs` pages via `before_timestamp`), so no in-repo consumer sends `offset > 0`. The behaviour change only affects external API clients — which is the intent.

## Findings

### Critical

**Functionality / docs contract — the OpenAPI spec now contradicts the implementation.**
`main` (6730d5d, merge of #1997) documents `/api/logs/search` as:
- `openapi/logs.yaml:287-297` — *"The CloudWatch provider … ignores `offset` … Paging past the first page there repeats results, so treat this endpoint as single-page"*
- `openapi/logs.yaml:325` — *"Ignored by the CloudWatch provider; see the description above"*
- `openapi/logs.yaml:326-332` — documented responses are `401`, `403`, `200`, `206`; there is **no `400`**.

After this change, a client that follows the published reference and sends `offset=10` gets a `400 INVALID_INPUT` from an endpoint whose spec says the parameter is silently ignored. `docs.json` serves that YAML as the public API reference, so the mismatch ships to users, and #1998 names removing this note as part of the fix. The branch was cut from `268d79d`, before #1997 merged, so this is easy to miss — but it has to be resolved before merge.

Fix: rebase on `main` and update `openapi/logs.yaml` — replace the "ignores offset" note with "the CloudWatch provider accepts only `offset=0` and returns `400 INVALID_INPUT` otherwise", update the `offset` parameter description, and add a `400` response entry (the repo's OpenAPI gate is `swagger-cli validate`, so a malformed edit will be caught there).

### Suggestion

**Functionality — the second half of #1998 is still open and now less visible.**
`backend/src/providers/logs/cloudwatch.provider.ts:528-531` still returns `total: mapped.length`, i.e. the size of the page just built. With `offset` pinned to 0, `paginatedResponse(res, logs, total, 0)` (`backend/src/api/routes/logs/index.routes.ts:136`) yields `200` + `Content-Range: 0-9/10` — the spec defines `200` as *"All matching log entries were returned"*, but the Insights query truncated at `limit`. Issue #1998's "stops early" failure mode therefore survives: a client asking for `limit=10` against 5,000 matching lines is told it has them all, and now has no offset escape hatch either. Not a blocker for option 3, but please cover it in the OpenAPI description you touch for the Critical item above, so the truncation is at least stated, and keep #1998 open (or file a follow-up) for the cursor work.

**Software engineering — validation is provider-local, so identical bad input still behaves differently per deployment.**
The route does `Number(offset)` with no validation (`backend/src/api/routes/logs/index.routes.ts:120,132`). With this PR, `offset=abc` → `NaN` → `400` on CloudWatch, while `LocalFileProvider` does `results.slice(NaN, NaN + limit)` → `slice(0, NaN)` → an empty page returned as `200` (`backend/src/providers/logs/local.provider.ts:253`); `offset=-10` similarly slices the last 10 rows there. Since the PR is explicitly about making the offset contract honest, consider validating `limit`/`offset` once in the route (finite, non-negative integer) so every provider agrees on malformed input, keeping the provider guard as a backstop.

**Software engineering — no test pins the `offset = 0` happy path.**
`backend/tests/unit/cloudwatch-provider-normalize.test.ts:330-338` only asserts rejection. A guard that over-triggers (`!offset` inverted, `offset !== undefined`, etc.) would break every CloudWatch search and the suite would stay green. One cheap line closes that: on the uninitialised `provider`, `searchLogs('error', undefined, 10, 0)` should reject with `LOGS_AWS_NOT_CONFIGURED` / `statusCode: 500` — proving execution got *past* the new guard. Same for the omitted-argument default.

### Information

- **Redundant condition** — `backend/src/providers/logs/cloudwatch.provider.ts:434`: `!Number.isFinite(offset) || offset !== 0` is equivalent to `offset !== 0`, since every non-finite number (`NaN`, `±Infinity`) already fails `!== 0`. The first clause documents intent but is dead; `offset !== 0` alone still passes all five parametrised cases.
- **Test file scope** — the new `describe` lands in `cloudwatch-provider-normalize.test.ts`, whose header comment scopes the file to the private `normalizeBody`/`parseRawLine` helpers. A `cloudwatch-provider-search.test.ts` (or a file rename) would keep that boundary clean. The `rejects.toMatchObject({ statusCode, code, message })` pattern itself matches existing convention (e.g. `webscraper-local.provider.test.ts:257-258`).
- **Error precedence changed** — the guard sits before the `cwLogGroup`/`cwClient` check, so an unconfigured deployment receiving `offset>0` now returns `400 INVALID_INPUT` instead of `500 LOGS_AWS_NOT_CONFIGURED`. Input errors first is the better ordering; just noting it's a behaviour change.
- **`AppError` supports `nextActions`** (28 uses in `backend/src`) — passing something like *"retry with offset=0 and narrow the time range for older entries"* would make the 400 actionable for clients that can't tell which provider answered.
- **Security** — no security-relevant changes. The route is admin-gated (`router.use(verifyAdmin)`, `index.routes.ts:12`), the change adds no new input reaching SQL/shell/HTTP, no dependency changes, and the new message leaks no secrets or PII. (The pre-existing regex interpolation of `q` into the Insights query at `cloudwatch.provider.ts:452` is untouched and out of scope here.)
- **Performance** — no impact, marginally positive: the guard fails before `StartQueryCommand`, so rejected requests no longer spend a CloudWatch Insights query plus up to 20×300 ms of polling.

## Verdict

`request_changes` — one Critical item: `openapi/logs.yaml` still tells clients CloudWatch ignores `offset` and documents no `400`. Update the spec (and ideally add the `offset=0` regression assertion) and this is good to go; the code change itself is small, correct, and faithful to the option the issue asked for.

**Comment by Gautam-aman** [comment]
@jwfing see if it is fine now?
